### PR TITLE
Replace hardcoded maxCalendarYear with modified spec validation

### LINKED ISSUE #9383: Support configuring maxCalendarYear for Schedules beyond 2100
**Is your feature request related to a problem? Please describe.**
The `maxCalendarYear` for Schedules is currently hardcoded to 2100. 
- This prevents users from scheduling workflows or timers that need to extend beyond that year, which can be a blocker for long-running or far-future scheduling use cases.
- or are expected to update Schedules once this limit is increased 

**Describe the solution you'd like**
Either increase the default `maxCalendarYear` limit, or make it configurable via `DynamicConfig` so that operators can adjust it without requiring a code change or redeployment.

**Describe alternatives you've considered**

**Additional context**
- this was a question from customer for payments

### PR DESCRIPTION
**Pull request by zivenrokhit**
## Decouple Spec Year Validation from Calendar Calculation Bound

Allow schedule specs to include years up to 9999, while keeping execution-time calendar search bounded at 2100 for safety.

## What changed?

- Added a separate spec validation limit: maxSpecYear = 9999.
- Kept maxCalendarYear = 2100, but clarified it is only a calculation safety bound during next-time lookup.
- Updated canonicalizeSpec and checkRanges to validate years against maxSpecYear instead of maxCalendarYear.
- Added TestSpecFarFutureYear to verify that a far-future year (for example, 2150) is accepted in the spec but resolves to zero-time when computation exceeds the 2100 bound.

## Why?

The previous approach used one limit for two different concerns:
- What input we allow users to express.
- How far the system is willing to search at runtime.

This change separates those concerns:
- Validation is permissive (up to 9999) so users can define far-future schedules.
- Execution remains protected (up to 2100) to avoid unbounded or expensive calendar calculations and reduce DoS risk.

Resolves #9383

## How did you test it?

- [x] built
- [x] run locally and tested manually
- [x] covered by existing tests
- [x] added new unit test(s)

## Potential risks

- Users may assume that if a far-future year is accepted, it will also produce an executable next time immediately; beyond 2100 it will not, and may return zero-time.
- The validation-vs-execution distinction may be non-obvious without clear documentation.
- If real execution beyond 2100 is needed later, raising maxCalendarYear will require careful performance and safety evaluation.

### CODE REVIEW

File: chasm/lib/scheduler/fx.go [hunk 1]
```diff
@@ -13,10 +14,14 @@ func Register(
 	return registry.Register(library)
 }
 
+func newSpecBuilder(dc *dynamicconfig.Collection) *legacyscheduler.SpecBuilder {
+	return legacyscheduler.NewSpecBuilder(dynamicconfig.SchedulerMaxCalendarYear.Get(dc)())
+}
```
[the remark this answers is not in the document]
Comment (zivenrokhit) [reply chasm/lib/scheduler/fx.go]: Addressed by adding boundcheck

**Review by dnr — COMMENTED** [review]
I don't think this is a good solution (and the suggestion in the original bug report is misguided):

There's no problem with limiting actual execution of schedules to > 50 years in the future: the server will be updated and redeployed many times before that. The only real problem is that the execution limit is used in spec validation, when there's no need for that at all. That is, specs should be allowed to reference years farther in the future if they want. The matching times won't be calculated that far in the future (this is DoS protection, basically), but as soon as an updated server is deployed, they will be.

The fix should be to just stop imposing a maximum year validation on specs.

**Comment by zivenrokhit** [comment]
Thanks for the feedback, I see what you are saying. I've scrapped the initial solution and relaxed the spec validation to allow for years beyond 2100, while keeping the match time DoS protection in place. I just want to highlight one possible edge case, if a user creates a schedule where the very next run isn't until after 2100, I believe the iterator will hit the DoS limit, fail to find a match, and immediately close the schedule as completed. Are we okay with this?

**Comment by chaptersix** [comment]
@zivenrokhit could you update the title and description?

**Comment by zivenrokhit** [comment]
> @zivenrokhit could you update the title and description?

Done

**Comment by chaptersix** [comment]
Please add functional tests covering the RPCs impacted by relaxing spec validation beyond year 2100:

**1. `ListScheduleMatchingTimes` with a far-future end time**

Create a weekly schedule, then call `ListScheduleMatchingTimes` with a start time of 2026-01-01 and an end time well beyond `maxCalendarYear`. Verify the RPC succeeds, returned times are correct (day of week, within `maxCalendarYear`), and the result count hits `maxListMatchingTimesCount`.

**2. `DescribeSchedule` with a far-future schedule**

Create a schedule with a spec that spans beyond `maxCalendarYear`, then call `DescribeSchedule`. Verify `FutureActionTimes` is populated and all entries are within `maxCalendarYear`.

**3. `DescribeSchedule` with a start year entirely beyond `maxCalendarYear`**

Create a schedule where all action times are beyond the calculation bound (e.g. `Year: "2150"`). Verify `CreateSchedule` succeeds but `DescribeSchedule` returns empty `FutureActionTimes`.

**4. `UpdateSchedule` to a far-future spec**

Create a schedule with a simple spec, then update it to a calendar spec referencing years beyond `maxCalendarYear`. Use `Calendar` string form so defaults are auto-filled. Verify the update succeeds and `DescribeSchedule` reflects the change with `FutureActionTimes` within `maxCalendarYear`. Note: for V1, the update is async so wrap describe assertions in `Eventually`.

Place these on `scheduleFunctionalSuiteBase` so they exercise both V1 and CHASM paths.


**Unit tests to add in `spec_test.go`:**

- Validate the `maxSpecYear` boundary: a spec with `Year: "9999"` should be accepted by `canonicalizeSpec`, while `Year: "10000"` should be rejected.
- A spec that straddles `MaxCalendarYear`: e.g. `Year: "2098-2150"` should return action times for 2098, 2099, 2100 and then stop (zero time after that).
- A spec spanning the full allowed range: `Year: "2000-9999"` (minCalendarYear to maxSpecYear) should be accepted, return action times starting from 2000, and stop computing at `MaxCalendarYear`.

**Comment by chaptersix** [comment]
> Please add functional tests covering the RPCs impacted by relaxing spec validation beyond year 2100:
> 
> **1. `ListScheduleMatchingTimes` with a far-future end time**
> 
> Create a weekly schedule, then call `ListScheduleMatchingTimes` with a start time of 2026-01-01 and an end time well beyond `maxCalendarYear`. Verify the RPC succeeds, returned times are correct (day of week, within `maxCalendarYear`), and the result count hits `maxListMatchingTimesCount`.
> 
> **2. `DescribeSchedule` with a far-future schedule**
> 
> Create a schedule with a spec that spans beyond `maxCalendarYear`, then call `DescribeSchedule`. Verify `FutureActionTimes` is populated and all entries are within `maxCalendarYear`.
> 
> **3. `DescribeSchedule` with a start year entirely beyond `maxCalendarYear`**
> 
> Create a schedule where all action times are beyond the calculation bound (e.g. `Year: "2150"`). Verify `CreateSchedule` succeeds but `DescribeSchedule` returns empty `FutureActionTimes`.
> 
> **4. `UpdateSchedule` to a far-future spec**
> 
> Create a schedule with a simple spec, then update it to a calendar spec referencing years beyond `maxCalendarYear`. Use `Calendar` string form so defaults are auto-filled. Verify the update succeeds and `DescribeSchedule` reflects the change with `FutureActionTimes` within `maxCalendarYear`. Note: for V1, the update is async so wrap describe assertions in `Eventually`.
> 
> Place these on `scheduleFunctionalSuiteBase` so they exercise both V1 and CHASM paths.
> 
> **Unit tests to add in `spec_test.go`:**
> 
> * Validate the `maxSpecYear` boundary: a spec with `Year: "9999"` should be accepted by `canonicalizeSpec`, while `Year: "10000"` should be rejected.
> * A spec that straddles `MaxCalendarYear`: e.g. `Year: "2098-2150"` should return action times for 2098, 2099, 2100 and then stop (zero time after that).
> * A spec spanning the full allowed range: `Year: "2000-9999"` (minCalendarYear to maxSpecYear) should be accepted, return action times starting from 2000, and stop computing at `MaxCalendarYear`.

done

File: service/worker/scheduler/spec_test.go [hunk 2]
```diff
@@ -494,3 +493,71 @@ func (s *specSuite) TestSpecJitterSeed() {
...
+	// Year 10000 (beyond maxSpecYear) should be rejected.
+	_, err = canonicalizeSpec(&schedulepb.ScheduleSpec{
+		StructuredCalendar: []*schedulepb.StructuredCalendarSpec{
+			{Year: []*schedulepb.Range{{Start: 10000}}},
+		},
+	})
+	s.Error(err)
+}
+
+func (s *specSuite) TestSpecYearStraddlesMaxCalendarYear() {
```
Comment (lina-temporal) [inline service/worker/scheduler/spec_test.go]: Think this is covered by `TestSpecFullYearRange`, I'd get rid of this one.

File: tests/schedule_test.go [hunk 3]
```diff
@@ -949,6 +953,249 @@ func testListScheduleMatchingTimes(t *testing.T, newContext contextFactory) {
...
+	s.Equal(time.Date(2026, 1, 8, 0, 0, 0, 0, time.UTC), firstTime)
+
+	lastTime := times[len(times)-1].AsTime()
+	s.LessOrEqual(lastTime.Year(), scheduler.MaxCalendarYear)
+}
+
+// testDescribeScheduleFarFutureActionTimes verifies that DescribeSchedule returns
+// FutureActionTimes within the MaxCalendarYear bound for a schedule that has no
+// end date and will run indefinitely.
+func testDescribeScheduleFarFutureActionTimes(t *testing.T, newContext contextFactory) {
```
Comment (lina-temporal) [inline tests/schedule_test.go]: I don't understand how this test is showing the "far future" (it doesn't appear to be really testing anything other than the immediate future), or really how it'd differentiate itself from any of the existing tests that validate `FutureActionTimes`.

File: tests/schedule_test.go [hunk 4]
```diff
@@ -949,6 +953,249 @@ func testListScheduleMatchingTimes(t *testing.T, newContext contextFactory) {
...
+		Namespace:  s.Namespace().String(),
+		ScheduleId: sid,
+		Schedule:   schedule,
+		Identity:   "test",
+		RequestId:  uuid.NewString(),
+	})
+	s.NoError(err)
+
+	// Update to a calendar spec that references a year beyond MaxCalendarYear.
+	// Use Calendar (string form) so defaults are auto-filled for unspecified fields.
```
Comment (lina-temporal) [inline tests/schedule_test.go]:
```suggestion
```

strike the comment imo

File: tests/schedule_test.go [hunk 5]
```diff
@@ -949,6 +953,249 @@ func testListScheduleMatchingTimes(t *testing.T, newContext contextFactory) {
...
+	schedule.Spec = &schedulepb.ScheduleSpec{
+		Calendar: []*schedulepb.CalendarSpec{
+			{
+				Year:       "2026-2217",
```
Comment (lina-temporal) [inline tests/schedule_test.go]: Something like `scheduler.MaxCalendarYear - 10` that tracks the constant would be good.

**Review by lina-temporal — COMMENTED** [review]
some small comments on tests, otherwise LGTM. Let's get @dnr's stamp as well.

File: tests/schedule_test.go (hunk 3 above)
Comment (chaptersix) [reply tests/schedule_test.go]: Added `TestDescribeScheduleFutureActionTimesAtBoundary` that should test what I was aiming at.

File: service/worker/scheduler/calendar.go [hunk 6]
```diff
@@ -30,9 +30,13 @@ type (
...
 	minCalendarYear = 2000
-	// maxCalendarYear is the latest year that will be recognized for calendar dates.
-	// If you're still using Temporal in 2100 please change this constant and rebuild.
-	maxCalendarYear = 2100
+	// MaxCalendarYear is the upper limit for calendar date computation. Specs can reference
+	// years beyond this limit, but actual calculation will stop here. Future server versions
+	// can safely increase this limit.
+	MaxCalendarYear = 2100
+	// maxSpecYear is the largest year allowed in a schedule spec. This is set to a very
+	// large value to effectively remove the restriction while still preventing absurd values.
```
Comment (dnr) [inline service/worker/scheduler/calendar.go]: the wording "remove this restriction" only makes sense if you knew there was a restriction before.. word the comment so it make sense for readers reading the current state of the code

File: tests/schedule_test.go [hunk 7]
```diff
@@ -951,6 +955,253 @@ func testListScheduleMatchingTimes(t *testing.T, newContext contextFactory) {
...
+	_, err = s.FrontendClient().UpdateSchedule(ctx, &workflowservice.UpdateScheduleRequest{
+		Namespace:  s.Namespace().String(),
+		ScheduleId: sid,
+		Schedule:   schedule,
+		Identity:   "test",
+		RequestId:  uuid.NewString(),
+	})
+	s.NoError(err)
+
+	s.Eventually(func() bool {
```
Comment (dnr) [inline tests/schedule_test.go]: use EventuallyWithT to make this more readable

File: service/worker/scheduler/calendar.go [hunk 8]
```diff
@@ -30,9 +30,13 @@ type (
 const (
 	// minCalendarYear is the smallest year that can appear in a calendar spec.
 	minCalendarYear = 2000
-	// maxCalendarYear is the latest year that will be recognized for calendar dates.
-	// If you're still using Temporal in 2100 please change this constant and rebuild.
-	maxCalendarYear = 2100
+	// MaxCalendarYear is the upper limit for calendar date computation. Specs can reference
+	// years beyond this limit, but actual calculation will stop here. Future server versions
+	// can safely increase this limit.
+	MaxCalendarYear = 2100
```
Comment (dnr) [inline service/worker/scheduler/calendar.go:36]: did we consider something like, stopping the calendar search at `ts + 100 years` instead having a hardcoded limit? that might make things more predictable all over

File: tests/schedule_test.go (hunk 7 above)
Comment (zivenrokhit) [reply tests/schedule_test.go]: done

File: service/worker/scheduler/calendar.go (hunk 6 above)
Comment (zivenrokhit) [reply service/worker/scheduler/calendar.go]: good call, I've reworded it so that it doesn't reference the old restriction

File: service/worker/scheduler/calendar.go (hunk 8 above)
Comment (zivenrokhit) [reply service/worker/scheduler/calendar.go:36]: I see the point about a relative limit, but I think sticking with a fixed year is the better move here. It keeps the behavior deterministic and much simpler to test since the boundary is consistent for every call. A relative limit adds another moving part that could introduce edge cases or inconsistencies
